server/main.js

```diff
@@ -46,6 +106,66 @@ function genidcode(){
 	return code;
 }
 
+function dedupNewFile(fileid) {
+	hashFile(`${FILES_DIRNAME}/${fileid}`, (digest, err) => {
+		if (err) {
+			console.log(`[dedup] Error hashing newly uploaded file: ${err}`);
+			return;
+		}
+
+		digest = digest.toString("base64");
+		const orig_fileid = hashmap.get(digest);
+		if (orig_fileid != null) {
+			// We've uploaded a file that already existed. Let's hard-link this
+			// file to the old one.
+			const oldpath = `${FILES_DIRNAME}/${orig_fileid}`;
+			const newpath = `${FILES_DIRNAME}/${fileid}`;
+
+			// We need to do this sequence of actions atomically, to ensure
+			// that we don't try to concurrently access the new file while
+			// we're in the process of replacing it by a link. Furthermore we
+			// need to be careful in case anything goes wrong.
+			try {
+				fs.renameSync(newpath, newpath + "-linkbak");
```

Comment:
Isn't it an idea to always write to an `-inprogress` file first on upload, then do a duplication check:
- rename it to the actual place if it doesn't exist or the original file is broken
- or link the original to the actual place if the original file exists and remove the `-inprogress` one